pkg/envoy/registry/services.go

```diff
@@ -176,9 +176,11 @@ func (k *KubeProxyServiceMapper) handleServiceDelete(svc *v1.Service) {
 			}
 			rem = append(rem, s)
 		}
+		k.cacheLock.RUnlock()
 		k.cacheLock.Lock()
 		k.servicesForCN[cn] = rem
 		k.cacheLock.Unlock()
+		k.cacheLock.RLock()
```

Comment:
Why can't we use sync.Map here?

Reply:
We could, though the docs generally recommend against it: https://golang.org/pkg/sync/#Map

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

Reply:
I think if all you need is protected access to the map, sync.Map should suffice. Do you see any reasons why this is not applicable here (I have not investigated the drawbacks of using sync.Map)?

Reply:
I haven't investigated that beyond those docs, and I'm not sure how closely this relates to the specific use cases sync.Map is designed for. I'll change it though if you think it's an improvement.

Reply:
The only reason to suggest using it is that one doesn't need to deal with locking/unlocking semantics as done now, which makes the code a bit easier to follow. Don't have a strong opinion on this.

Reply:
> The main annoying thing is dealing with what happens when you look up a CN that doesn't exist. With the plain map, it's just a slice with a value of nil. With sync.Map it's still nil, but panics if you assert it to be a slice.

Could you share some code snippets to compare this? Trying to understand if this is any concern. I would rather prefer to avoid a bunch of locks, unlocks for conciseness if there is no additional drawback here, so I'd like to see what the concern is regarding using assertions.

Reply:
https://play.golang.org/p/AP3D_DqfCFS

Not that it's impossible to work around, but doing that would be just as verbose as the locking.

Reply:
```
	sval, ok = sm.Load("notkey")
	fmt.Printf("sm[\"notkey\"], ok = %#v, %v\n", sval.([]string), ok)
```

I don't agree the problem is the same here. The above lookup on an invalid key can be easily circumvented by checking if the key exists (derived from `ok`). I would rather prefer that we always handle invalid lookups explicitly rather than relying on the internal implementation of the library to return a nil value for further computation.

Reply:
I pushed a change to use sync.Map. Let me know what you think.

Reply:
@nojnhuh I don't have a strong opinion on this particular change here wrt locking, but want to ensure this discussion leads to less ambiguity in future changes. Hope we can follow similar patterns in other areas of the code rather than being inconsistent without a good reason.